Microsoft.Identity.Web/Client/TokenCacheProviders/InMemory/MSALAppMemoryTokenCacheProvider.cs

```diff
@@ -0,0 +1,147 @@
+
+namespace Microsoft.Identity.Web.Client.TokenCacheProviders
+{
+    /// <summary>
+    /// An implementation of token cache for Confidential clients backed by MemoryCache.
+    /// MemoryCache is useful in Api scenarios where there is no HttpContext to cache data.
+    /// </summary>
+    /// <seealso cref="https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/wiki/token-cache-serialization"/>
+    public class MSALAppMemoryTokenCacheProvider : IMSALAppTokenCacheProvider
+    {
+        /// <summary>
+        /// The application cache key
+        /// </summary>
+        internal string AppCacheId;
+
+        /// <summary>
+        /// The backing MemoryCache instance
+        /// </summary>
+        internal IMemoryCache memoryCache;
+
+        /// <summary>
+        /// The duration till the tokens are kept in memory cache. In production, a higher value , upto 90 days is recommended.
+        /// </summary>
+        private readonly DateTimeOffset cacheDuration = DateTimeOffset.Now.AddHours(12);
+
+        /// <summary>
+        /// The internal handle to the client's instance of the Cache
+        /// </summary>
+        private ITokenCache ApptokenCache;
+
+        private IConfiguration Configuration;
+
+        public MSALAppMemoryTokenCacheProvider(IConfiguration configuration, IMemoryCache cache)
+        {
+            this.Configuration = configuration;
+            this.memoryCache = cache;
+        }
+
+        /// <summary>Initializes this instance of TokenCacheProvider with essentials to initialize themselves.</summary>
+        /// <param name="tokenCache">The token cache instance of MSAL application</param>
+        /// <param name="httpcontext">The Httpcontext whose Session will be used for caching.This is required by some providers.</param>
+        public void Initialize(ITokenCache tokenCache, HttpContext httpcontext)
+        {
+            AzureADOptions azureAdOptions = new AzureADOptions();
+            this.Configuration.Bind("AzureAD", azureAdOptions);
+
+            this.AppCacheId = azureAdOptions.ClientId + "_AppTokenCache";
+
+            this.ApptokenCache = tokenCache;
+            this.ApptokenCache.SetBeforeAccess(this.AppTokenCacheBeforeAccessNotification);
+            this.ApptokenCache.SetAfterAccess(this.AppTokenCacheAfterAccessNotification);
+            this.ApptokenCache.SetBeforeWrite(this.AppTokenCacheBeforeWriteNotification);
+
+            this.LoadAppTokenCacheFromMemory();
+        }
+
+        /// <summary>
+        /// if you want to ensure that no concurrent write take place, use this notification to place a lock on the entry
+        /// </summary>
+        /// <param name="args">Contains parameters used by the MSAL call accessing the cache.</param>
+        private void AppTokenCacheBeforeWriteNotification(TokenCacheNotificationArgs args)
+        {
+            // Since we are using a MemoryCache ,whose methods are threads safe, we need not to do anything in this handler.
```

Comment:
Then why do we subscribe to the event at all?

Reply:
For the comment i provided here. I want the users to know why we didn't need to do anything in this event handler.

---
In reply to: [267388310](https://github.com/Azure-Samples/active-directory-aspnetcore-webapp-openidconnect-v2/pull/65#discussion_r267388310) [](ancestors = 267388310)